File: src/interpreter/expression/variable.ts
import { NameNotFound } from '../error'
import * as cal from 'calcium-js'

export class Variable implements cal.Assignable {
  constructor(public readonly name: string) {}

  assign(value: cal.Any, env: cal.Environment) {
    env.context.register(this.name, value)
  }

  evaluate(env: cal.Environment): cal.Any {
    const value = env.context.lookUp(this.name)
    if (value === undefined) {
      throw new NameNotFound(this.name)
    }
    return value
  }
}

File: lib/interpreter/pseudo/expression/variable.dart
import 'package:calcium_app/interpreter/expression.dart';
import 'package:calcium_app/interpreter/types.dart';
import 'package:calcium_app/interpreter/runtime_data.dart';

class Variable extends Reference {
  final String name;

  Variable(this.name);

  @override
  void assign(Any value, RuntimeData data) {
    data.context.register(name, value);
  }

  @override
  Any evaluate(RuntimeData data) {
    return data.context.lookUp(name);
  }
}
